server/cluster/cluster.go

```diff
@@ -1082,6 +1082,72 @@ func (c *RaftCluster) processRegionHeartbeat(region *core.RegionInfo) error {
 	return nil
 }
 
+func (c *RaftCluster) processRegionSplit(regions []*metapb.Region) []error {
+	if err := c.checkSplitRegions(regions); err != nil {
+		return []error{err}
+	}
+	total := len(regions) - 1
+	regions[0], regions[total] = regions[total], regions[0]
+	leaderStoreID := uint64(0)
+	if r := c.core.GetRegion(regions[0].GetId()); r != nil {
+		leaderStoreID = r.GetLeader().GetStoreId()
+	}
+	errList := make([]error, 0, total)
+	for _, region := range regions {
+		// It should update rightmost because it can override by left.
+		if len(region.GetPeers()) == 0 {
+			errList = append(errList, errors.New(fmt.Sprintf("region:%d has no peer", region.GetId())))
+			continue
+		}
+		// region split initiator store will be leader with a high probability
+		leader := region.Peers[0]
+		if leaderStoreID > 0 {
+			for _, peer := range region.GetPeers() {
+				if peer.GetStoreId() == leaderStoreID {
+					leader = peer
+					break
+				}
+			}
+		}
+		region := core.NewRegionInfo(region, leader)
+		overlaps, perr := c.core.AtomicCheckAndPutRegion(region)
+		if perr != nil {
+			errList = append(errList, perr)
+			continue
+		}
+		for _, item := range overlaps {
```

Comment:
This loop also appears in `processRegionHeartbeat`, how about extracting it?

Reply:
fixed